### PR TITLE
Format Chainer E2E transformer forward (fixed)

### PR DESCRIPTION
**Pull request by Fhrozen**
In order to make more similar the scripts between pytorch and chainer, this PR simplifies the script in forward function (E2E_transformer).
Some functions have been moved into the encoder and decoder depending its required
This PR does not intent to check docs into the transformer so I would recommend to check and merge this one first and then #1017

Replace of #1028

### CODE REVIEW

File: espnet/nets/chainer_backend/ctc.py [hunk 1]
```diff
@@ -127,6 +129,35 @@ def __call__(self, hs, ys):
 
         return self.loss
 
+    def forward_from_transformer(self, hs, ys):
+        """Core function of the Warp-CTC layer.
+
+        Args:
+            hs (iterable of chainer.Variable | N-dimention array): Input variable from encoder.
+            ys (iterable of chainer.Variable | N-dimension array): Input variable of decoder.
```
Comment (ShigekiKarita) [inline espnet/nets/chainer_backend/ctc.py]: I do not think the `ys` can be chainer.Variable. see https://github.com/jheymann85/chainer_ctc/blob/master/chainer_ctc/warpctc.py#L34

File: espnet/nets/chainer_backend/ctc.py [hunk 2]
```diff
@@ -127,6 +129,35 @@ def __call__(self, hs, ys):
 
         return self.loss
 
+    def forward_from_transformer(self, hs, ys):
```
Comment (ShigekiKarita) [inline espnet/nets/chainer_backend/ctc.py]: If this method is only for Transformer, I would like to find this in Transformer class

File: espnet/nets/chainer_backend/e2e_asr_transformer.py [hunk 3]
```diff
@@ -169,80 +155,38 @@ def forward(self, xs, ilens, ys_pad, calculate_attentions=False):
...
-                cer_ctc = self.error_calculator(ys_hat, ys_pad_cpu, is_ctc=True)
+                cer_ctc = self.error_calculator(ys_hat, ys_pad, is_ctc=True)
 
-        # Compute cer/wer
-        with chainer.no_backprop_mode():
-            y_hats = ys.reshape((batch, length, -1))
-            y_hats = chainer.backends.cuda.to_cpu(F.argmax(y_hats, axis=2).data)
+        # 3. Decoder
+        ys = self.decoder(ys_pad, xs, x_mask)
+        if calculate_attentions:
```
Comment (ShigekiKarita) [inline espnet/nets/chainer_backend/e2e_asr_transformer.py]: instead of return nothing, I suggest to create new method like `def calculate_attentions(self, xs, xs_mask, ys_pad)`

File: espnet/nets/chainer_backend/transformer/label_smoothing_loss.py [hunk 4]
```diff
@@ -19,9 +19,17 @@ def __init__(self, smoothing, n_target_vocab, normalize_length=False):
...
-
-    def forward(self, concat_logit_block, t_block, batch, length):
+        self.ignore_id = ignore_id
+
+    def forward(self, ys_block, ys_pad, eos):
+        xp = self.xp
+        eos = np.array([eos], 'i')
+        ys_out = [np.concatenate([y, eos], axis=0) for y in ys_pad]
+        ys_out = F.pad_sequence(ys_out, padding=-1)
+        t_block = chainer.Variable(xp.array(ys_out.data))
```
Comment (ShigekiKarita) [inline espnet/nets/chainer_backend/transformer/label_smoothing_loss.py]: use `with chainer.no_backprop_mode():`

File: espnet/nets/chainer_backend/ctc.py [hunk 5]
```diff
@@ -90,6 +90,8 @@ class WarpCTC(chainer.Chain):
 
     def __init__(self, odim, eprojs, dropout_rate):
```
Comment (ShigekiKarita) [inline espnet/nets/chainer_backend/ctc.py]: maybe you need to change `__call__` to `forward` in this class

File: espnet/nets/chainer_backend/transformer/decoder.py [hunk 6]
```diff
@@ -48,27 +50,49 @@ def __init__(self, odim, args, initialW=None, initial_bias=None):
...
-            chainer.Chain: Decoder layer.
+    def forward(self, ys_pad, source, x_mask):
+        """Forward decoder
 
+        :param xp.array e: input token ids, int64 (batch, maxlen_out)
+        :param xp.array yy_mask: input token mask, uint8  (batch, maxlen_out)
+        :param xp.array source: encoded memory, float32  (batch, maxlen_in, feat)
+        :param xp.array xy_mask: encoded memory mask, uint8  (batch, maxlen_in)
+        :return e: decoded token score before softmax (batch, maxlen_out, token)
+        :rtype: chainer.Variable
```
Comment (ShigekiKarita) [inline espnet/nets/chainer_backend/transformer/decoder.py]: use Google style doc

Comment (Fhrozen) [reply espnet/nets/chainer_backend/transformer/decoder.py]: I am leaving this for #1017

File: espnet/nets/chainer_backend/transformer/label_smoothing_loss.py [hunk 7]
```diff
@@ -44,4 +52,5 @@ def forward(self, concat_logit_block, t_block, batch, length):
             label_smoothing = F.sum(label_smoothing) / normalizer
             loss = self.confidence * loss + self.smoothing * label_smoothing
 
-        return loss
+        acc = F.accuracy(concat_logit_block, concat_t_block, ignore_label=self.ignore_id)
+        return loss, acc
```
[the remark this answers is not in the document]
Comment (Fhrozen) [reply espnet/nets/chainer_backend/transformer/label_smoothing_loss.py]: Ok, it is fine is it is stored inside of a self.acc in the label and then called from the e2e.forward?

File: espnet/nets/chainer_backend/transformer/decoder.py (hunk 6 above)
Comment (ShigekiKarita) [reply espnet/nets/chainer_backend/transformer/decoder.py]: OK

File: espnet/nets/chainer_backend/transformer/label_smoothing_loss.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ShigekiKarita) [reply espnet/nets/chainer_backend/transformer/label_smoothing_loss.py]: No. I do not want `LabelSmoothingLoss` to do too many things except for the loss calculation. I suggest:
- concat `eos` to `t_block` outside
- calculate acc outside

[the remark this answers is not in the document]
Comment (Fhrozen) [reply espnet/nets/chainer_backend/transformer/label_smoothing_loss.py]: done, pls check it

**Comment by sw005320** [comment]
I changed the target branch of this PR from the master to v.0.6.0 since it includes several refactoring.

File: .gitignore [hunk 8]
```diff
@@ -4,6 +4,7 @@
 \#*\#
 .\#*
 *DS_Store
+out.txt
```
Comment (sw005320) [inline .gitignore]: Do you need this?

File: .gitignore [hunk 9]
```diff
@@ -21,6 +22,7 @@ shellcheck*
 check_shellcheck*
 test_spm.vocab
 test_spm.model
+custom*
```
Comment (sw005320) [inline .gitignore]: Do you need this?

File: .gitignore [hunk 10]
```diff
@@ -36,6 +38,7 @@ egs/*/*/tensorboard
 egs/*/*/wav*
 
 # tools related
+tools/ideep*
```
Comment (sw005320) [inline .gitignore]: Do you need this?

File: espnet/nets/chainer_backend/transformer/subsampling.py [hunk 11]
```diff
@@ -15,32 +15,29 @@ class Conv2dSubsampling(chainer.Chain):
...
-                                initialW=initialW(scale=stvd),
-                                initial_bias=initial_bias(scale=stvd))
-        self.add_link('conv.2', layer)
-        stvd = 1. / np.sqrt(dims)
-        layer = L.Linear(idim, dims, initialW=initialW(scale=stvd),
-                         initial_bias=initial_bias(scale=stvd))
-        self.add_link('out.0', layer)
         self.dropout = dropout
         with self.init_scope():
+            n = 1 * 3 * 3
```
Comment (sw005320) [inline espnet/nets/chainer_backend/transformer/subsampling.py]: Can you add some comments about `n = 1 * 3 * 3`?

File: .gitignore (hunk 8 above)
Comment (Fhrozen) [reply .gitignore]: Pytest generates this file, so maybe

Comment (sw005320) [reply .gitignore]: I see. Then, let's keep this.

File: .gitignore (hunk 9 above)
Comment (Fhrozen) [reply .gitignore]: I was thinking about any custom training/decoding setup used for egs so they can be ignore, something like custom_training_my_rnn.yaml

File: .gitignore (hunk 10 above)
Comment (Fhrozen) [reply .gitignore]: I was trying to speed up chainer for cpu so I installed it directly, i can remove it

File: espnet/nets/chainer_backend/transformer/subsampling.py (hunk 11 above)
Comment (Fhrozen) [reply espnet/nets/chainer_backend/transformer/subsampling.py]: ok

File: .gitignore (hunk 9 above)
Comment (sw005320) [reply .gitignore]: ok, but it may not be started from `custom*`. So, we may not need this.

**Comment by Fhrozen** [comment]
The main difference between the ctc for transformer and the rnn is because the target (`ys`) is already a list of arrays located in the cpu, while in rnn routine the target is a list of variables located in cpu. If the target of rnn becomes a list of cpu arrays then it should no be a problem.

**Comment by sw005320** [comment]
> The main difference between the ctc for transformer and the rnn is because the target (`ys`) is already a list of arrays located in the cpu, while in rnn routine the target is a list of variables located in cpu. If the target of rnn becomes a list of cpu arrays then it should no be a problem.

OK.
Can you add this comment to the source? 
You may add it as a TODO to unify this module across the PyTorch and Chainer backends, and we may care this in later PRs if needed.

**Comment by sw005320** [comment]
Thanks, @Fhrozen!
